File: src/pytra/runtime/std/re.py
from __future__ import annotations


def _is_space(ch: str) -> bool:
    return ch == " " or ch == "\t" or ch == "\n" or ch == "\r" or ch == "\f" or ch == "\v"


def sub(pattern: str, repl: str, text: str) -> str:
    if pattern == r"\s+":
        out: list[str] = []
        i = 0
        n = len(text)
        while i < n:
            ch = text[i]
            if _is_space(ch):
                while i < n and _is_space(text[i]):
                    i += 1
                out.append(repl)
                continue
            out.append(ch)
            i += 1
        return "".join(out)
    return text.replace(pattern, repl)

File: src/runtime/cpp/pytra/std/re.cpp
#include "runtime/cpp/pytra/built_in/py_runtime.h"


namespace pytra::std::re {

    
    bool _is_space(const str& ch) {
        return (ch == " ") || (ch == "\t") || (ch == "\n") || (ch == "\r") || (ch == "") || (ch == "");
    }
    
    str sub(const str& pattern, const str& repl, const str& text) {
        if (pattern == "\\s+") {
            list<str> out = list<str>{};
            int64 i = 0;
            int64 n = py_len(text);
            while (i < n) {
                str ch = text[i];
                if (_is_space(ch)) {
                    while ((i < n) && (_is_space(text[i]))) {
                        i++;
                    }
                    out.append(str(repl));
                    continue;
                }
                out.append(str(ch));
                i++;
            }
            return py_to_string(str("").join(out));
        }
        return py_to_string(py_replace(text, pattern, repl));
    }
    
}  // namespace pytra::std::re
